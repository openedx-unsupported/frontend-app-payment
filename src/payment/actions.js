import { utils } from '../common';

const { AsyncActionType } = utils;

export const FETCH_BASKET = new AsyncActionType('PAYMENT', 'FETCH_BASKET');


export const fetchBasket = () => ({
  type: FETCH_BASKET.BASE,
  payload: {},
});

export const fetchBasketBegin = () => ({
  type: FETCH_BASKET.BEGIN,
});

export const fetchBasketSuccess = result => ({
  type: FETCH_BASKET.SUCCESS,
  payload: result,
});

export const fetchBasketFailure = () => ({
  type: FETCH_BASKET.FAILURE,
});
