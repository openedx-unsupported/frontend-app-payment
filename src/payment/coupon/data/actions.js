import { AsyncActionType } from '../../../common/utils';

export const ADD_COUPON = new AsyncActionType('PAYMENT', 'ADD_COUPON');
export const REMOVE_COUPON = new AsyncActionType('PAYMENT', 'REMOVE_COUPON');
export const UPDATE_COUPON_DRAFT = 'PAYMENT__UPDATE_COUPON_DRAFT';

// ADD COUPON

export const addCoupon = code => ({
  type: ADD_COUPON.BASE,
  payload: {
    code,
  },
});

export const addCouponBegin = () => ({
  type: ADD_COUPON.BEGIN,
});

export const addCouponSuccess = (voucherId, code, benefit) => ({
  type: ADD_COUPON.SUCCESS,
  payload: {
    benefit,
    code,
    voucherId,
  },
});

export const addCouponFailure = errorCode => ({
  type: ADD_COUPON.FAILURE,
  payload: {
    errorCode,
  },
});

// REMOVE COUPON

export const removeCoupon = voucherId => ({
  type: REMOVE_COUPON.BASE,
  payload: {
    voucherId,
  },
});

export const removeCouponBegin = () => ({
  type: REMOVE_COUPON.BEGIN,
});

export const removeCouponSuccess = result => ({
  type: REMOVE_COUPON.SUCCESS,
  payload: {
    result,
  },
});

export const removeCouponFailure = errorCode => ({
  type: REMOVE_COUPON.FAILURE,
  payload: {
    errorCode,
  },
});

// UPDATE COUPON DRAFT

export const updateCouponDraft = code => ({
  type: UPDATE_COUPON_DRAFT,
  payload: {
    code,
  },
});
