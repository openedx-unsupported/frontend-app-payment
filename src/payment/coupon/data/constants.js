export const PERCENTAGE_BENEFIT = 'Percentage';
export const ABSOLUTE_BENEFIT = 'Absolute';
