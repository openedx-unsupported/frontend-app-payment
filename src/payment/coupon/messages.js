import { defineMessages } from '@edx/frontend-i18n';

const messages = defineMessages({
  'payment.coupon.label': {
    id: 'payment.coupon.label',
    defaultMessage: 'Add coupon code (optional)',
    description: 'Label for the add coupon form',
  },
  'payment.coupon.submit': {
    id: 'payment.coupon.submit',
    defaultMessage: 'Apply',
    description: 'Submit button for the add coupon form',
  },
  'payment.coupon.remove': {
    id: 'payment.coupon.remove',
    defaultMessage: 'Remove',
    description: 'Submit button to remove a coupon',
  },
  'payment.coupon.benefit_value': {
    id: 'payment.coupon.benefit_value',
    defaultMessage: 'Coupon {code} applied for {value} off',
    description: 'The description of a coupon for a percentage off the total price. Percentage symbol is included in {value}',
  },
  'payment.coupon.benefit.default': {
    id: 'payment.coupon.benefit.default',
    defaultMessage: 'Coupon {code} applied',
    description: 'The description of a coupon without any specific monetary value off.',
  },
});

export default messages;
