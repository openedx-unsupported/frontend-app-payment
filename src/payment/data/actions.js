import { createRoutine } from 'redux-saga-routines';

// Routines are action + action creator pairs in a series.
// Actions adhere to the flux standard action format.
// Routines by default are in the form of:
//
// Action                |   Action Creator
// -----------------------------------------------
// fetchBasket.TRIGGER   |   fetchBasket()
// fetchBasket.SUCCESS   |   fetchBasket.success()
// fetchBasket.FAILURE   |   fetchBasket.failure()
// fetchBasket.FULFILL   |   fetchBasket.fulfill()
//
// Created with redux-saga-routines
export const fetchCaptureKey = createRoutine('FETCH_CAPTURE_KEY');
export const submitPayment = createRoutine('SUBMIT_PAYMENT');
export const fetchBasket = createRoutine('FETCH_BASKET');
export const addCoupon = createRoutine('ADD_COUPON');
export const removeCoupon = createRoutine('REMOVE_COUPON');
export const updateQuantity = createRoutine('UPDATE_QUANTITY');

// Actions and their action creators
export const BASKET_DATA_RECEIVED = 'BASKET_DATA_RECEIVED';

export const basketDataReceived = basket => ({
  type: BASKET_DATA_RECEIVED,
  payload: basket,
});

export const BASKET_PROCESSING = 'BASKET_PROCESSING';

export const basketProcessing = isProcessing => ({
  type: BASKET_PROCESSING,
  payload: isProcessing,
});

export const CAPTURE_KEY_PROCESSING = 'CAPTURE_KEY_PROCESSING';

export const captureKeyProcessing = isProcessing => ({
  type: CAPTURE_KEY_PROCESSING,
  payload: isProcessing,
});

export const MICROFORM_STATUS = 'MICROFORM_STATUS';

export const microformStatus = status => ({
  type: MICROFORM_STATUS,
  payload: status,
});

export const CAPTURE_KEY_START_TIMEOUT = 'CAPTURE_KEY_START_TIMEOUT';

export const captureKeyStartTimeout = () => ({
  type: CAPTURE_KEY_START_TIMEOUT,
});
export const CAPTURE_KEY_DATA_RECEIVED = 'CAPTURE_KEY_DATA_RECEIVED';

export const captureKeyDataReceived = captureKey => ({
  type: CAPTURE_KEY_DATA_RECEIVED,
  payload: captureKey,
});
