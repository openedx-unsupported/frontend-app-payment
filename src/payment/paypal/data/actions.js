import { utils } from '../../../common';

const { AsyncActionType } = utils;

export const SUBMIT_PAYMENT_PAYPAL = new AsyncActionType('PAYMENT', 'SUBMIT_PAYMENT_PAYPAL');

export const submitPaymentPayPal = () => ({
  type: SUBMIT_PAYMENT_PAYPAL.BASE,
});

export const submitPaymentPayPalBegin = () => ({
  type: SUBMIT_PAYMENT_PAYPAL.BEGIN,
});

export const submitPaymentPayPalFailure = () => ({
  type: SUBMIT_PAYMENT_PAYPAL.FAILURE,
});
