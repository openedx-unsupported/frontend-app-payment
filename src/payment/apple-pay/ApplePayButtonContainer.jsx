import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl, intlShape } from '@edx/frontend-i18n';

import { basketSelector } from '../data/selectors';
import { addMessage, clearMessages, MESSAGE_TYPES } from '../../feedback';
import ApplePayButton from './ApplePayButton';
import { redirectToReceipt } from './service';

import messages from './ApplePay.messages';

function ApplePayButtonContainer(props) {
  const handleMerchantValidationFailure = useCallback(() => {
    props.clearMessages();
    props.addMessage(
      'apple-pay-failure',
      props.intl.formatMessage(messages['payment.apple.pay.merchant.validation.failure']),
      null,
      MESSAGE_TYPES.WARNING,
    );
  });

  const handlePaymentAuthorizationFailure = useCallback(() => {
    props.clearMessages();
    props.addMessage(
      'apple-pay-failure',
      props.intl.formatMessage(messages['payment.apple.pay.authorization.failure']),
      null,
      MESSAGE_TYPES.ERROR,
    );
  });

  return (
    <ApplePayButton
      className={props.className}
      disabled={props.disabled}
      totalAmount={props.orderTotal}
      onPaymentComplete={redirectToReceipt}
      onMerchantValidationFailure={handleMerchantValidationFailure}
      onPaymentAuthorizationFailure={handlePaymentAuthorizationFailure}
      title={props.intl.formatMessage(messages['payment.apple.pay.pay.with.apple.pay'])}
      lang={props.intl.locale}
    />
  );
}

ApplePayButtonContainer.propTypes = {
  orderTotal: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  addMessage: PropTypes.func.isRequired,
  clearMessages: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
};

ApplePayButtonContainer.defaultProps = {
  orderTotal: undefined,
  className: undefined,
  disabled: false,
};

export default connect(basketSelector, {
  addMessage,
  clearMessages,
})(injectIntl(ApplePayButtonContainer));
