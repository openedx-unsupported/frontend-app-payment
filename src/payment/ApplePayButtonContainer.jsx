import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { sendTrackEvent } from '@edx/frontend-analytics';

import { ApplePayButton } from './payment-methods/apple-pay';
import { submitPayment } from './data/actions';

function ApplePayButtonContainer(props) {
  const handleClick = useCallback(() => {
    // TO DO: after event parity, track data should be
    // sent only if the payment is processed, not on click
    // Check for PayPal and Free Basket as well
    sendTrackEvent(
      'edx.bi.ecommerce.basket.payment_selected',
      {
        type: 'click',
        category: 'checkout',
        paymentMethod: 'Apple Pay',
      },
    );

    props.submitPayment({ method: 'apple-pay' });
  });

  return (
    <ApplePayButton
      className={`payment-method-button ${props.className}`}
      disabled={props.disabled}
      onClick={handleClick}
    />
  );
}

ApplePayButtonContainer.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  submitPayment: PropTypes.func.isRequired,
};

ApplePayButtonContainer.defaultProps = {
  className: undefined,
  disabled: false,
};

const mapStateToProps = state => ({
  disabled: state.payment.basket.submitting,
});

export default connect(mapStateToProps, { submitPayment })(ApplePayButtonContainer);
