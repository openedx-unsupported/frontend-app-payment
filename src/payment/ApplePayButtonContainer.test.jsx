/* eslint-disable global-require */
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { IntlProvider, configure as configureI18n } from '@edx/frontend-i18n';
import configureMockStore from 'redux-mock-store';

import { configuration } from '../environment';
import { submitPayment } from './data/actions';
import messages from '../i18n';
import ApplePayButtonContainer from './ApplePayButtonContainer';

const defaultStore = require('./__mocks__/defaultState.mockStore.js');


jest.mock('@edx/frontend-analytics', () => ({
  sendTrackEvent: jest.fn(),
}));
import { sendTrackEvent } from '@edx/frontend-analytics'; // eslint-disable-line import/first

// Mock language cookie
Object.defineProperty(global.document, 'cookie', {
  writable: true,
  value: `${configuration.LANGUAGE_PREFERENCE_COOKIE_NAME}=en`,
});

configureI18n(configuration, messages);
const mockStore = configureMockStore();
const applePaySession = { begin: jest.fn() };
global.ApplePaySession = jest.fn().mockImplementation(() => applePaySession);
global.ApplePaySession.canMakePayments = () => true;

describe('<ApplePayButtonContainer />', () => {
  it('should render properly', () => {
    const tree = renderer
      .create((
        <IntlProvider locale="en">
          <Provider store={mockStore(defaultStore)}>
            <ApplePayButtonContainer />
          </Provider>
        </IntlProvider>
      ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should be apple pay payment submission', () => {
    const store = mockStore(defaultStore);
    const wrapper = mount((
      <IntlProvider locale="en">
        <Provider store={store}>
          <ApplePayButtonContainer />
        </Provider>
      </IntlProvider>
    ));
    wrapper.find('ApplePayButtonContainer').simulate('click');

    // Track an event
    expect(sendTrackEvent).toHaveBeenCalled();

    // Dispatch the submit payment action
    const actions = store.getActions();
    expect(actions).toEqual([
      submitPayment({ method: 'apple-pay' }),
    ]);
  });
});
