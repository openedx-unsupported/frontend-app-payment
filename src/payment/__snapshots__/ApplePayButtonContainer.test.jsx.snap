// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ApplePayButtonContainer /> should render properly 1`] = `
<button
  className="apple-pay-button payment-method-button undefined"
  disabled={false}
  id="applePayBtn"
  lang="en"
  onClick={[Function]}
  title="Pay with Apple Pay"
/>
`;
